### ISSUE TITLE
Support per window rendering for native GUIs

### ISSUE THREAD

**Issue by rkannan82** [neovim/neovim#2464]
In order to build a GUI client that does its own window management, we need Neovim to provide changes per window. It seems like today the server assumes the client has one big window/canvas and it tells what text to put where. This is not ideal if we want to create native windows. For e.g., let's say we have neovim.app (GUI for OSX) and user creates a vertical split to show the navigation tree on the left side. Today, the window edges are created using symbols like "|" because that's what the server is sending. Correct me if I am wrong. If we want to draw the window border ourselves using OSX frames, then we really need to know the text to place inside each window. Please comment on the feasibility of this request.

Here is the reference issue I had raised in neovim.app: https://github.com/rogual/neovim-dot-app/issues/73

**Comment by kopischke** [comment]
:+1: to that. Right now, all a Neovim GUI can provide is a totally pointless menu, slightly saner key code processing and some fancy window chrome, all at the expense of having to implement its own configuration mechanism. Truly decoupling the editor from the UI will require us to stop treating clients as one big cell based canvas.

**Comment by tarruda** [comment]
I'm aware of the problem and have plans to address it after the first release(even mentioned in the new bountysource campaign). Some other related discussion: https://github.com/neovim/neovim/issues/2224#issuecomment-84730773

**Comment by fwalch** [comment]
> I'm aware of the problem and have plans to address it after the first release

Or even before the first release? :grin:  https://tarruda.github.io/articles/neovim-smart-ui-protocol/

**Comment by tarruda** [comment]
> Or even before the first release? :grin: https://tarruda.github.io/articles/neovim-smart-ui-protocol/

Yes. It may seem like this will hurt neovim stability and hold a first release even further, but as the post explains it will only be active when a certain command-line flag is passed, being implemented in a way that what we currently have is not affected.

**Comment by hoschi** [comment]
Is there a issue or a repo (for the Typescript proof of concept) I can watch the state of smart ui protocol? I didn't find something on this issue tracker.

**Comment by tarruda** [comment]
> Is there a issue or a repo (for the Typescript proof of concept) I can watch the state of smart ui protocol? I didn't find something on this issue tracker.

Not yet, but I may push something next week

**Comment by hoschi** [comment]
Please post a short update when this happens, I would love to see more
about this!

Thiago de Arruda <notifications@github.com> schrieb am Do., 25. Juni 2015
15:22:

**Comment by Philipp-M** [comment]
Is there any progress on this?

**Comment by justinmk** [comment]
@Philipp-M  #4432 enabled arbitrary rendering of the popup menu. "wildmenu" could be next candidate for that type of capability. 

The request in this issue is a much bigger change. Any progress on it will be noted here.

**Comment by teto** [comment]
A first attempt can be found here https://github.com/neovim/neovim/pull/5686
